src: more automatic memory management in node_crypto.cc

addaleax opened the pull request:
Prefer custom smart pointers fitted to the OpenSSL data structures
over more manual memory management and lots of `goto`s.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

addaleax commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14462/

tniessen commented on src/node_crypto.cc:
```diff
@@ -586,24 +595,26 @@ int SSL_CTX_use_certificate_chain(SSL_CTX* ctx,
 
       // Find issuer
-      if (*issuer != nullptr || X509_check_issued(ca, x) != X509_V_OK)
+      if (!*issuer || X509_check_issued(ca, x.get()) != X509_V_OK)
```
I might be missing something here, but doesn't this negate the check?

addaleax replied:
@tniessen Yes, good catch. I don’t think we have tests for this at the moment. :/

addaleax commented:
CI²: https://ci.nodejs.org/job/node-test-commit/17974/

bnoordhuis commented on src/node_crypto.cc:
```diff
@@ -4674,38 +4550,21 @@ class PBKDF2Request : public AsyncWrap {
-                int iter,
-                int keylen)
+                MallocedBuffer<char>&& pass,
+                MallocedBuffer<char>&& salt,
```
Coincidence: I was working yesterday on simplifying this to something that uses `std::vector<char>`. 

(Working on adding scrypt support and DRYing the PBKDF2 and RandomBytes code in the process.)

bnoordhuis commented on src/node_crypto.h:
```diff
@@ -73,6 +73,32 @@ struct MarkPopErrorOnReturn {
   ~MarkPopErrorOnReturn() { ERR_pop_to_mark(); }
 };
 
+template<typename T, void (*function)(T*)>
```
Style: space before `<`.

bnoordhuis commented on src/node_crypto.h:
```diff
@@ -73,6 +73,32 @@ struct MarkPopErrorOnReturn {
+template<typename T, void (*function)(T*)>
+struct FunctionDeleter {
+  void operator()(T* pointer) const { function(pointer); }
+  typedef std::unique_ptr<T, FunctionDeleter> ptr;
```
Can you name this `Ptr`, `Pointer`, `UniquePointer` or whatever to make it clearer it's a type?

bnoordhuis commented on src/node_crypto.cc:
```diff
@@ -543,24 +553,23 @@ void SecureContext::SetKey(const FunctionCallbackInfo<Value>& args) {
-  return ret;
+  DeleteFnPtr<X509_STORE_CTX, X509_STORE_CTX_free> store_ctx(
+      X509_STORE_CTX_new());
+  return store_ctx != nullptr &&
```
I suppose this works because of boolean coercion but it might be clearer to write it as `!!store_ctx` or `store_ctx == true` or `store_ctx.get() != nullptr`.

addaleax commented:
Fixed tests (not sure why I didn’t see the failures before) and addressed @bnoordhuis’ first round of review nits.

CI: https://ci.nodejs.org/job/node-test-pull-request/14469/

tniessen commented on src/node_crypto.h:
```diff
@@ -478,19 +479,16 @@ class SignBase : public BaseObject {
   void CheckThrow(Error error);
 
-  EVP_MD_CTX* mdctx_;
+  DeleteFnPtr<EVP_MD_CTX, EVP_MD_CTX_free> mdctx_;
```
Is there a reason not to reuse `EVPMDPointer`?

addaleax replied:
@tniessen thanks for catching, updated!

addaleax commented:
CI: https://ci.nodejs.org/job/node-test-commit/18082/

Would anybody in particular like me to wait with landing until they have time to review? I know it’s a large commit, and not urgent, so I really don’t want to push this to land early.

bnoordhuis commented on src/node_crypto.cc:
```diff
@@ -4249,32 +4137,32 @@ void DiffieHellman::ComputeSecret(const FunctionCallbackInfo<Value>& args) {
+      0));
+
+  int dataSize = DH_size(diffieHellman->dh_.get());
   char* data = Malloc(dataSize);
```
Maybe use automatic memory management for this as well (and rename `dataSize` to `data_size` while you're here)?  Just a suggestion.

addaleax replied:
Good idea – I’ve switched this to `MallocedBuffer`, too (that takes care of `dataSize` anyway)

bnoordhuis commented on src/node_crypto.cc:
```diff
@@ -4428,34 +4315,31 @@ void ECDH::GenerateKeys(const FunctionCallbackInfo<Value>& args) {
 
-  pub = EC_POINT_new(group);
+  ECPointPointer pub(EC_POINT_new(group));
   if (pub == nullptr) {
```
`!pub`, also on line 4366.

bnoordhuis commented on src/node_crypto.cc:
```diff
@@ -5320,33 +5136,32 @@ void ConvertKey(const FunctionCallbackInfo<Value>& args) {
 
-  EC_GROUP* group = EC_GROUP_new_by_curve_name(nid);
+  ECGroupPointer group(
+    EC_GROUP_new_by_curve_name(nid));
```
4 space indent.

bnoordhuis commented on src/util.h:
```diff
@@ -438,6 +437,37 @@ struct OnScopeLeave {
+  }
+  MallocedBuffer& operator=(MallocedBuffer&& other) {
+    this->~MallocedBuffer();
+    return *new (this)MallocedBuffer(other);
```
Style: `new(this)`

Don't know if it matters here but use of placement new sometimes inhibits  compiler optimizations.

addaleax replied:
> Don't know if it matters here but use of placement new sometimes inhibits compiler optimizations.

I don’t think we actually use it at this point, so this is just here for completeness

bnoordhuis commented on src/util.h:
```diff
@@ -438,6 +437,37 @@ struct OnScopeLeave {
+  MallocedBuffer& operator=(const MallocedBuffer&) = delete;
+};
+
+}  // namespace node
```
Is this class necessary?  A `std::unique_ptr` with a custom deleter would be sufficient too, wouldn't it?

addaleax replied:
@bnoordhuis Yes, the difference being that this also tracks the length of the buffer by itself. If you feel strongly about it I can switch to splitting up into two variables again

bnoordhuis replied:
No strong feelings.

addaleax commented:
CI: https://ci.nodejs.org/job/node-test-commit/18166/

tniessen commented:
Rerunning Windows: https://ci.nodejs.org/job/node-test-commit-windows-fanned/17684/

addaleax commented:
Rebased

CI before landing: https://ci.nodejs.org/job/node-test-commit/18218/

addaleax commented:
Another day, another merge conflict, another rebase. :)

CI: https://ci.nodejs.org/job/node-test-commit/18232/

addaleax commented:
Landed in d7cba76856e80108edb47b2e4c50a2c5fe5e1427

MylesBorins commented:
Should this be backported to `v1.x-staging`? If yes please follow [the guide](https://github.com/nodejs/node/blob/master/doc/guides/backporting-to-release-lines.md) and raise a backport PR, if not let me know or add the `dont-land-on` label.
